File: network/src/main/java/com/mars/network/gsonfactory/BaseResponseBodyConverter.java
package com.mars.network.gsonfactory;

import android.text.TextUtils;

import com.google.gson.Gson;
import com.google.gson.TypeAdapter;

import com.mars.network.base.BaseException;
import com.ymars.poj.base.Constant;
import com.ymars.poj.comutils.KtStringUtils;
import com.ymars.poj.comutils.LogTools;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;

import okhttp3.ResponseBody;
import retrofit2.Converter;

/**
 * @author Mars
 * 时间：2019/08/23
 * 描述：请求返回基类转化工厂
 */

public class BaseResponseBodyConverter<T> implements Converter<ResponseBody, T> {
    private Gson gson;
    private TypeAdapter<T> adapter;

    BaseResponseBodyConverter(Gson gson, TypeAdapter<T> adapter) {
        this.gson = gson;
        this.adapter = adapter;
    }

    @Override
    public T convert(ResponseBody value) throws IOException {
        String jsonString = value.string();
        LogTools.Companion.i("value", KtStringUtils.Companion.isNotNullString(jsonString) ? jsonString : "is null");
        try {
            JSONObject object = new JSONObject(jsonString);
            int status = object.getInt(Constant.code);
            if (status != 0) {
                if (status == -1001) {
                     LogTools.Companion.i("当前状态", "当前状态未登录");
                }
                String msg = object.getString(Constant.msg);
                if (TextUtils.isEmpty(msg)) {
                    msg = String.format("[%s]", status);
                }
                //异常处理
                throw new BaseException(msg, status);
            } else {
                String data = object.getString(Constant.data);
                if (KtStringUtils.Companion.isNotNullString(data)) {
                     LogTools.Companion.i("Data", data);
                    return adapter.fromJson(data);
                } else {
                     LogTools.Companion.i("Data", "data is null");
                    return adapter.fromJson("{}");
                }
            }

        } catch (JSONException e) {
            e.printStackTrace();
            //数据解析异常
             LogTools.Companion.i(this.getClass().getSimpleName(), e.getMessage());
            throw new BaseException(BaseException.PARSE_ERROR_MSG, BaseException.PARSE_ERROR);
        } finally {
            value.close();
        }
    }
}

File: network/src/main/java/com/mars/network/gsonfactory/BaseResponseBodyConverter.kt
package com.mars.network.gsonfactory

import com.google.gson.Gson
import com.google.gson.TypeAdapter
import com.mars.network.base.BaseException
import com.mars.network.base.BaseReponseModel
import com.ymars.poj.base.Constant
import com.ymars.poj.comutils.LogTools
import okhttp3.ResponseBody
import org.json.JSONObject
import retrofit2.Converter
import java.io.IOException
import java.nio.charset.Charset

class BaseResponseBodyConverter<T>(
    private val gson: Gson,
    private val adapter: TypeAdapter<T>
) : Converter<ResponseBody, T> {
    @Throws(IOException::class)
    override fun convert(value: ResponseBody): T? {
        val jsonString = value.string()
        LogTools.i(
            "BaseResponseBodyConverter",
            jsonString
        )
      val resp: BaseReponseModel<T>? =
          gson.fromJson(jsonString, BaseReponseModel::class.java) as BaseReponseModel<T>?;
        LogTools.i(
            "BaseResponseBodyConverter",
            resp.toString()
        )
        LogTools.i(
            "BaseResponseBodyConverter",
            resp!!.res.toString()
        )
        if(resp!!.code==0){
            val jsonObject = JSONObject(jsonString)
            val data: String = jsonObject.getString(Constant.data)
            return adapter.fromJson(data)
        }else{
            throw BaseException("数据解析错误", 0xE1);
        }
    }
    companion object {
        private val UTF_8 = Charset.forName("UTF-8")
    }
}